### PR TITLE
Datasync cli

### PR DESCRIPTION
**Pull request by Eunovo**
### Description

Early stage of work on graphback plugin to generate datasync files. Currently;
- json schema for models are generated
- datasync config file initializing all the models is generared

##### Checklist

- [x] `npm test` passes
- [x] `npm run build` works
- [x] tests are included
- [x] documentation is changed or added
- [x] integration with sample app

### CODE REVIEW

File: packages/offix-datasync-codegen/package.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "@offix/datasync-plugin",
+  "version": "0.1.0",
+  "description": "Graphback plugin for offix-datasync ",
```
Comment (wtrocki) [inline packages/offix-datasync-codegen/package.json:4]: Client

File: packages/offix-datasync-codegen/declarations.d.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+declare module 'graphql-metadata';
```
Comment (wtrocki) [inline packages/offix-datasync-codegen/declarations.d.ts:1]: Is there issue with typings for metadata?

File: packages/offix-datasync-codegen/src/OffixDataSyncConfig.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Config for OffixDataSync plugin
+ */
+export interface IOffixDataSyncPluginConfig {
+    /**
+     * The output dir path.
+     * The json schema will be written to ${outputDir}/schema.json
+     * The datasync config will be written to ${outputDir}/config.ts
+     */
+    outputDir?: string;
```
Comment (wtrocki) [inline packages/offix-datasync-codegen/src/OffixDataSyncConfig.ts:10]: Rename to model output dir

File: packages/offix-datasync-codegen/tests/__snapshots__/DataSyncClientPlugin.test.ts.snap [hunk 4]
```diff
@@ -0,0 +1,71 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Test datasync config 1`] = `
```
Comment (wtrocki) [inline packages/offix-datasync-codegen/tests/__snapshots__/DataSyncClientPlugin.test.ts.snap:3]: Not sure if we need to generate config or it should be just part of the docs.

File: packages/offix-datasync-codegen/declarations.d.ts (hunk 2 above)
Comment (Eunovo) [reply packages/offix-datasync-codegen/declarations.d.ts:1]: Yes

**Comment by wtrocki** [comment]
Amazing work. 

## Crazy idea for UX

command should ask couple questions - what is your desired output config etc and then create GraphQL.config file. For that we will need to extend graphback cli :( so it might be too much for this PR but it might be worth to chat about it

File: packages/offix-datasync-codegen/tests/__snapshots__/DataSyncClientPlugin.test.ts.snap (hunk 4 above)
Comment (Eunovo) [reply packages/offix-datasync-codegen/tests/__snapshots__/DataSyncClientPlugin.test.ts.snap:3]: We can generate but still leave it editable to the user. Like generate boilerplate code. To ensure that we don't overwrite the user's work, we can add a config option to the plugin to skip generating datasync config

File: packages/offix-datasync-codegen/declarations.d.ts (hunk 2 above)
Comment (wtrocki) [reply packages/offix-datasync-codegen/declarations.d.ts:1]: Cc @craicoverflow Let's fix them then

File: packages/offix-datasync-codegen/tests/__snapshots__/DataSyncClientPlugin.test.ts.snap (hunk 4 above)
Comment (wtrocki) [reply packages/offix-datasync-codegen/tests/__snapshots__/DataSyncClientPlugin.test.ts.snap:3]: Trick is to not generate anything that should be editable, but ok to leave it

**Comment by Eunovo** [comment]
@wtrocki  @kingsleyzissou I have added graphback to sample app and some cli documentation. The docs are far behind and need explanation on the schema fields but this PR should be mergeable. Also, I couldn't add much to the schema.json because I'm not sure how most of the new fields should work but It should be easy to edit what is generated.
Please review and merge is no changes are required.

**Comment by Eunovo** [comment]
@wtrocki It won't be easy to test the cli integration without publishing the plugin. I'm experiencing problems using the linked packages on my end due to problems related to multiple installed graphql modules.

**Comment by kingsleyzissou** [comment]
@Eunovo will review now. As for the docs, I think things are changing quite quickly, so I would say maybe we can nail down the architecture and our MVPs and focus on the docs when we have something a bit more solid. @wtrocki WDYT?

**Comment by wtrocki** [comment]
> @wtrocki WDYT?

See https://github.com/aerogear/offix/issues/659

@kingsleyzissou let's publish package for eunovo and add myself as maintainer (aerogear-ci) as well

File: docs/ref-datastore-getting-started.md [hunk 5]
```diff
@@ -101,3 +101,36 @@ The `DataStore` constructor takes a schema version parameter(defaults to 1).
...
+You can import models from `config.ts` and start coding! 
+
+### Installing the plugin
+
+For npm
+`npm install offix-datasync-client-plugin`
+or yarn
+`yarn add offix-datasync-client-plugin`
+
+Add the plugin to your `.graphqlrc.yml`
```
Comment (wtrocki) [inline docs/ref-datastore-getting-started.md:120]: you will not have .graphqlrc.yml. That is why we need small wrapper that will create it with some sensible defaults for our plugin

Comment (wtrocki) [reply docs/ref-datastore-getting-started.md:120]: This can be separate issue

File: examples/react-datastore/.graphqlrc.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+schema: './src/schema.graphql'
+extensions:
+  graphback:
+    # path to data mode file(s)
+    model: './src/model/runtime.graphql'
+    plugins:
+      offix-datasync-client-plugin:
```
Comment (wtrocki) [inline examples/react-datastore/.graphqlrc.yml:7]: I think it is best for us to also build schema for developers to see/share etc.

Comment (Eunovo) [reply examples/react-datastore/.graphqlrc.yml:7]: Can we take out the schema reference? I tried removing it and got some errors

Comment (wtrocki) [reply examples/react-datastore/.graphqlrc.yml:7]: This goes against this file
This file is build to describe schema. 
You can get schema from link if you wish.... 

But generally we should generate visible schema for users - drop schema plugin explicitly and dump schema somewhere.

**Pull request merged**